Re-factor to fix high CPU usage on windows

tsg opened the pull request:
This decouples getting the basic state information from getting the more
detailed information.  The reason is that filtering only needs the name, but
the code was getting all details and then apply filtering. Getting the command
line is expensive on Windows, so that contributed to #1460.

What made it worse is that due to filtering, the command line cache was
invalidated on each run, which explains why topbeat was consuming more CPU when
filtering out processes than when not.

Fixes #1460.

This also pre-compiles the regexps, which brings some slight CPU usage improvements
and uncovers compilation errors earlier.

tsg commented:
Don't merge yet, please, I need to test if it all looks good under Windows.

tsg commented on topbeat/beater/topbeat.go:
```diff
@@ -115,7 +115,10 @@ func (t *Topbeat) Setup(b *beat.Beat) error {
 	var err error
 
-	t.procStats.InitProcStats()
+	err = t.procStats.InitProcStats()
```
@ruflin this might conflict with #1081, let's see which PR gets merged first so we can updated the other.

tsg commented on topbeat/system/process.go:
```diff
@@ -211,20 +235,22 @@ func (procStats *ProcStats) GetProcStats() ([]common.MapStr, error) {
-			last, ok := procStats.ProcsMap[process.Pid]
-			if ok {
-				procStats.ProcsMap[process.Pid] = process
-			}
```
@monicasarbu I wasn't sure about the reasoning behind this `if ok` branch. Seems like the ProcsMap will get overwritten anyway.

monicasarbu replied:
@tsg I think you can simply remove the entire `if` as ProcsMap is updated in "procStats.ProcMap = newProcs" below.

tsg commented:
Tested on Windows and OS X, looks good and the CPU usage dropped significantly on Windows.

ruflin commented:
@tsg #1081 and #1566 were both merged with changes in topbeat. Rebase on top of it probably makes sense even though there does not seem to be a conflict. `make fmt` and one topbeat test seem to be failing.

tsg commented:
@ruflin there was one slight change in the API (InitProcesses now can return an error). I rebased and added a commit to add the error check to Metricbeat. Please have a look.

The test and go vet should also be fixed now.

ruflin commented:
LGTM. Waiting for Travis...

ruflin commented:
Merging without Travis as Travis seems to be stuck ...
